src/Biryuchkov/Task5/main.cpp

```diff
@@ -0,0 +1,157 @@
+constexpr auto HEIGHT = 960;
+constexpr auto FOV = M_PI/3.;
+constexpr auto MAX_REFLECTIONS = 4;
+constexpr auto BACKGROUND_COLOR = QVector3D(1.f, 1.0f, 0.7f);
+
+
+QVector3D mix(const QVector3D &a, const QVector3D &b, const float k){
+    return k * a + (1 - k) * b;
+}
+
+
+QVector3D reflect(const QVector3D &incident, const QVector3D &normal) {
+    return incident - 2.f * normal * (QVector3D::dotProduct(incident, normal));
+}
+
+
+QVector3D refract(const QVector3D &incident, const QVector3D &normal, const float n_1, const float n_2=1.f) {
+    auto cosine_1 = - std::max(-1.f, std::min(1.f, QVector3D::dotProduct(incident, normal)));
+    if (cosine_1 < 0.f) return refract(incident, -normal, n_2, n_1);
+    auto ratio = n_2 / n_1;
+    auto cosine_2 = sqrtf(1 - ratio * ratio * (1 - cosine_1 * cosine_1));
+    return  ratio * incident + (ratio * cosine_1 - cosine_2) * normal;
+}
+
+
+bool scene_intersect(const Ray &ray, const std::vector<Sphere> &spheres, const std::vector<Surface> &planes, QVector3D &hit, QVector3D &normal, Material &material) {
+    auto spheres_dist = std::numeric_limits<float>::max();
+    for (const auto & sphere : spheres) {
+        auto dist_i = std::numeric_limits<float>::max();
+        if (sphere.ray_intersect(ray, dist_i) && dist_i < spheres_dist) {
+            spheres_dist = dist_i;
+            hit = ray.origin + ray.direction * dist_i;
+            normal = (hit - sphere.center).normalized();
+            if(sphere.material.isTexture()) {
+                int u = (sphere.material.img.width()-1) * (1 + atan2(normal.z(), normal.x()) / M_PI) * 0.5f;
+                int v = (sphere.material.img.height()-1) * acosf(normal.y()) / M_PI;
+                QColor tmp = sphere.material.img.pixel(u, v);
+                material.diffuseColor = QVector3D(tmp.redF(), tmp.greenF(), tmp.blueF());
+            }else {
+                material = sphere.material;
+            }
+        }
+    }
+    float planes_dist = std::numeric_limits<float>::max();
+    for (const auto & plane : planes) {
+        auto dist_i = std::numeric_limits<float>::max();
+        if (plane.ray_intersect(ray, dist_i) && dist_i < spheres_dist) {
+            planes_dist = dist_i;
+            hit = ray.origin + ray.direction * dist_i;
+            normal = plane.normal;
+            material.diffuseColor = (int(.5*hit.x()+10.) + int(.5*hit.z())) & 1 ? QVector3D(.9f, .9f, .9f) : QVector3D(.0f, .0f, .0f);
+        }
+    }
+    return std::min(spheres_dist, planes_dist)<1000.f;
+}
+
+
+QVector3D cast_ray(const Ray &ray, const std::vector<Sphere> &spheres, const std::vector<Surface> &planes, const std::vector<Light> &lights, size_t depth=0){
+    QVector3D point;
+    QVector3D normal;
+    Material material;
+
+    if (depth > MAX_REFLECTIONS || !scene_intersect(ray, spheres, planes, point, normal, material)){
+        return BACKGROUND_COLOR;
+    }
+    QVector3D reflect_direction = reflect(ray.direction, normal).normalized();

```

Comment:
```suggestion
    const auto reflect_direction = reflect(ray.direction, normal).normalized();

```

Если компилятор может вывести тип, то пусть он это делает за вас.